Add an --exec flag to pass decrypted secrets via environment variables to a child process

ancat opened the pull request:
Hi there, I'm looking at using sops for an infrastructure project. My current use-case looks something like this:

1. Create an AWS EC2 instance with an IAM instance profile that lets it talk to KMS and is pre-loaded with secrets encrypted by sops
2. At boot, the instance executes sops to decrypt the file and writes it to disk.
3. A daemon that loads this file is started. It uses credentials in this file to connect to a remote database, operate on HMACs, etc.

This daemon can alternatively pull secret values out of its environment, which is where this PR comes in. I would much prefer to pass the credential to this daemon using the environment for a few reasons:
- it's possible to pass the credential without it ever touching the disk in plaintext
- it separates the secret encrypted values from the rest of the config. this is similar to what the `--unencrypted-suffix` does, but we can have configurations that are sourced from the same config file but different environments (eg production vs staging) can substitute in different secrets

My current technique to address this involves using process substitution in a bash shim. This works but isn't the best solution (not least of all because it doesn't work properly on OS X) The shim does something like this:

```
# this is equivalent to `source /some/env/file/on/disk`
# bash silently substitutes the command with something like /dev/pts/40
# which is a pipe connected to the stdout of the process
source <(sops -d /etc/secrets.env)

# /some/config.yml contains regular content, and the secrets are implicitly
# pulled from the environment because they're not found in the config file
/usr/bin/some-program --config /some/config.yml
```

My PR adds an `--exec` flag that does just this. In short, after decryption, dotenv output is added to a one time use environment that is made available to a single child process. This child process can then retrieve the plaintext secret without having to retrieve it from the disk.

I think it'll provide a lot of benefit to folks using dotenv output types. dotenv files are ultimately meant to be loaded into a process' environment, and this does that for you while making it easy to do securely. Writing out secret values to disk and then loading them into your environment opens up room for more risk such as improper permissions on output files, leaking the (now sensitive) environment to more processes than the one that actually needs it, etc. Also I skimmed the issues and saw I'm not the only one looking for something like this (#482, #449)

Some examples:

```
$ ./sops -d test.env
# This is an example file.
hmac_key=abc123
database_password=0th348w0hf3w9ru3pjro

$ ./sops --exec 'bash' -d test.env
bash-3.2$ echo $database_password
0th348w0hf3w9ru3pjro

$ ./external_processor.py
Couldn't find the database password. Make sure to pass it in as an environment variable.

$ ./sops --exec './external_processor.py' -d test.env
Found it! Here: 0th348w0hf3w9ru3pjro

$ cat external_processor.py
#!/usr/bin/python

import os

secret_key = os.getenv("database_password")
if not secret_key:
    print "Couldn't find the database password. Make sure to pass it in as an environment variable."
else:
    print "Found it! Here: {}".format(secret_key)
```

Let me know if you have any questions or request any changes. If this approach seems sane, I can go ahead and write some tests as well.

Thank you for reading!

autrilla commented:
Thanks for the writeup and contribution. It does look like there are several people interested in this, but I'm not sure this is the right way to implement it. I have a couple of concerns:

- I don't think it's a good idea to do this only for one storage format, but I also don't see a good way to support this for all formats, as some provide nested structures. Having it only work for some formats is confusing and makes the code more complex, imo.
- I don't like adding yet another flag to the `sops` command -- it's already very cluttered, and when calling `sops --help` it is difficult to tell which flags can be used when. I think a `sops exec` subcommand would be better.

jvehent commented:
At a mimum, I wouldn't call this exec, but something like `--source-env`
since it effectively sources an env file.

On Sun, Aug 4, 2019, 04:16 Adrian Utrilla <notifications@github.com> wrote:

> Thanks for the writeup and contribution. It does look like there are
> several people interested in this, but I'm not sure this is the right way
> to implement it. I have a couple of concerns:
>
>    - I don't think it's a good idea to do this only for one storage
>    format, but I also don't see a good way to support this for all formats, as
>    some provide nested structures. Having it only work for some formats is
>    confusing and makes the code more complex, imo.
>    - I don't like adding yet another flag to the sops command -- it's
>    already very cluttered, and when calling sops --help it is difficult
>    to tell which flags can be used when. I think a sops exec subcommand
>    would be better.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/sops/pull/504?email_source=notifications&email_token=AADFPAOK7BCTFEV27UGYSKLQC2F5RA5CNFSM4IJEPRF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3P43CQ#issuecomment-517983626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADFPAJZA4KYNNRPGRORKY3QC2F5RANCNFSM4IJEPRFQ>
> .
>

ancat commented:
Thank you for the feedback!

@autrilla, it's only the output type that matters here. Your input can be json, yaml, etc but if you use dotenv chances are you'll eventually want them loaded in your env anyway. The same can't be said for the other types, which is why it doesn't make sense for json/yaml/binary.

RE your second point, I agree. I initially prototyped a subcommand but decided against it since all the subcommands deal with operations that are operations distinct from the main command that handles encryption and decryption. There's a lot of clutter, but I think that could be handled by breaking encryption/decryption/key management out into their own subcommands as well, not so much making an entire subcommand just for this. I can take another stab at it if you think this is the right way to go.

@jvehent Yep, it does that but I think the more important part is the Exec that happens right after the sourcing. Otherwise there's no one to pass the new environment variables to.

autrilla commented:
> @autrilla, it's only the output type that matters here. Your input can be json, yaml, etc but if you use dotenv chances are you'll eventually want them loaded in your env anyway. The same can't be said for the other types, which is why it doesn't make sense for json/yaml/binary.

.ini could be used for this as well, I guess. And maybe non-nested YAML/JSON. We could make it work with all formats, and exit with an error if the tree contains nested data.

> There's a lot of clutter, but I think that could be handled by breaking encryption/decryption/key management out into their own subcommands as well, not so much making an entire subcommand just for this. I can take another stab at it if you think this is the right way to go.

That is indeed something I believe we should do longer-term. We've (or at least I have 😉) put newer functionality onto subcommands to prevent making things worse. So I think it's better to make this a subcommand.

In my opinion, setting environment variables from SOPS could be a good thing to do from an exec command -- but I'd like to see a 'generic' `sops exec` that works for all storage formats, and could for example have two modes that I can t think of right now:

- Set environment variables: this should work for all files that don't contain nested structures
- Decrypt a file into a temporary file, and delete it after the called command terminates.

It is probably best to split those into two subcommands, `sops exec-with-file` and `sops exec-with-env` or something along those lines.

ancat commented:
Sounds good! I'll update the PR once I've got something together.

ancat commented:
Hey @autrilla, I took your suggestions and moved exec features into its own subcommand and merged the temporary file idea + the original env idea into a single one. I thought they were close enough that it might make sense to just leave em in the same command. Summary of the changes:

```
$ sops exec -h
NAME:
   sops exec - execute a command with decrypted values inserted into the environment

USAGE:
   sops exec [command options] [arguments...]

OPTIONS:
   --placeholder              pass the decrypted contents as a file to the command instead of through the environment
   --background               background the process and don't wait for it to complete
   --enable-local-keyservice  use local key service
   --keyservice value         Specify the key services to use in addition to the local one. Can be specified more than once. Syntax: protocol://address. Example: tcp://myserver.com:5000
```

This subcommand has two approaches to passing secrets to a new process:
1. via environment variables (default)
This decrypts the input file, inserts them into a new environment, and launches a new process with that environment. Not too different from before.
2. via a fifo (`--placeholder`)
This one might be a little weird. Basically this creates a fifo to act as a temporary file and replaces all instances of `{}` with its path. It's similar to find(1)'s `-exec` flag (executes a command per each file it finds) I went with a fifo because it behaves like a file but has the added bonus of only being readable once (by the child process). Basically, the child process reads the file (fifo), parses, runs, etc all without the plaintext ever having to touch the disk. This can address issues that would otherwise leave decrypted files around (process crashes before we can clean up, something else ends up copying the file, the child process leaves the file descriptor open so the file is never truly deleted, etc)

I also added a `--background` flag since in some of my usecases I didn't need the process to be interactive or wait for it etc.

Some example usage:
```
$ sops exec out.json 'echo hello from sops: $AWS_ACCESS_KEY_ID'
hello from sops: AKIAIOSFODNN7EXAMPLE

$ sops exec --placeholder out.json 'echo your file: {} ; cat {}'
your file: /tmp/.sops015630600/tmp-file
{
        "database_password": "jf48t9wfw094gf4nhdf023r",
        "AWS_ACCESS_KEY_ID": "AKIAIOSFODNN7EXAMPLE",
        "AWS_SECRET_KEY": "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
}
```

autrilla commented:
I'm not sure how I feel about the FIFO. It does have some added advantages as you mention, but I imagine there's processes that might want to read the file more than one. I've ran into this in the past with e.g. `ssh`. I think it might be best to split the 'env' and 'file' behavior into two, and then have a flag for the file to be a regular temporary file instead of a FIFO for those who need that. The advantages of a pipe are good enough that I feel that should be the default.

The `--background` flag is probably okay to have, although most shells can achieve similar behavior.

Before we merge this, we'll need documentation in the README, but it's probably best to wait until the implementation is set :)

autrilla commented:
This looks great to me. Thanks for working on it @ancat! I'm very happy with the implementation now, so the only thing missing to merge this would be documentation on the README.

ancat commented:
@autrilla oops, I didn't see your comment! I snuck in a setuid flag (--user) and was wondering what you thought of that. I could take it or leave it, but in many cases I found myself needing to drop privileges after decrypting (basically read secrets as root, drop privilege to some daemon user, start process). Let me know what you think!

ancat commented:
also looks like the build failed for what looks like reasons unrelated to the most recent change:

```
E: Package 'rpm' has no installation candidate
apt-get.diagnostics
apt-get install failed
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install rpm ruby python3 unzip" failed and exited with 100 during .
```

autrilla commented:
> @autrilla oops, I didn't see your comment! I snuck in a setuid flag (--user) and was wondering what you thought of that. I could take it or leave it, but in many cases I found myself needing to drop privileges after decrypting (basically read secrets as root, drop privilege to some daemon user, start process). Let me know what you think!
> 
> ```
> $ sudo sops exec-file --user daemon out.json 'TMPFILE={} sh'
> Password:
> sh-3.2$ id
> uid=1(daemon) gid=1(daemon) groups=1(daemon),12(everyone),61(localaccounts)
> sh-3.2$ cat $TMPFILE
> {
>         "database_password": "jf48t9wfw094gf4nhdf023r",
>         "AWS_ACCESS_KEY_ID": "AKIAIOSFODNN7EXAMPLE",
>         "AWS_SECRET_KEY": "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
> }
> 
> $ sudo sops exec-env --user daemon out.json 'echo $AWS_ACCESS_KEY_ID / $AWS_SECRET_KEY / `whoami`'
> AKIAIOSFODNN7EXAMPLE / wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY / daemon
> ```

That seems like it could be useful. Adding a short note to the README about what it can be used for would be helpful too!

autrilla commented:
@ancat reran the build, it fails for another reason now ;)

ancat commented:
@autrilla, looks good now I think! I'll get started on the updated documentation in a bit. :~)

ancat commented:
Of course, and thank you for maintaining sops :~)
